backend/drm/drm.c

```diff
@@ -208,10 +209,41 @@ static bool init_planes(struct wlr_drm_backend *drm) {
 		 * tracking of them for DRM lease clients.
 		 *
 		 * possible_crtcs is a bitmask of crtcs, where each bit is an
-		 * index into drmModeRes.crtcs. So if bit 0 is set (ffs starts
-		 * counting from 1), crtc 0 is possible.
-		 */
-		int crtc_bit = ffs(plane->possible_crtcs) - 1;
+		 * index into drmModeRes.crtcs. So if bit i is set crtc i is
+		 * possible.
+                 *
+		 * The loop below is selecting the first possible CRTC without a
+		 * associated plane of the considered type or, as a fallback,
+		 * the last possible CRTC.
+                 */
+
+		int crtc_bit ;
+		int num_crtcs = (int)drm->num_crtcs ;
+		assert(num_crtcs<=64);
+		uint64_t bits = plane->possible_crtcs ;
+		// Some drm implementations (e.g. vc4 on raspberry pi) may set more
+		// bits than there are CRTCs so the unused upper bits must be cleared
+		// to insure a proper detection of the last possible CRTC.
+		bits &= (((uint64_t)1u)<<num_crtcs) - 1;
+		for (crtc_bit = 0 ; crtc_bit < num_crtcs ; crtc_bit++) {
+			if (bits & 1) {
+				if (bits==1) {
+					break; // The last CRTC
```

Comment:
Why is this necessary? `crtc_bit < num_crtcs` should end the loop at the correct time, right?

Reply:
No. If the for loop terminates via crtc_bit < num_crtcs  then this is an error case (no valid CRTC for the plane) and the assert is triggered.  That break is there to catch the last bit set in possible_crtcs. 
For example, consider the following hypothetical case:
 - 4 crtcs in 2 groups of 2 
 - 8 overlay planes. the first half have possible_crtcs = 0b0011 while the second half have 0b1100.    
The break is there to insure that the supplementary overlay planes (those that would not pass the test crtc->num_overlays>0) are assigned to a valid crtc. 

More generally, if you plan to accept the pull request then it is probably better to consider the the alternative implementation I proposed in https://github.com/swaywm/wlroots/pull/2333#discussion_r456779587.

Reply:
>alternative implementation

Yeah that one looks better